Implement BIP 340-342 validation (Schnorr/taproot/tapscript)

pinheadmz opened the pull request:
This branch is modeled after https://github.com/bitcoin/bitcoin/pull/17977 which is the current work-in-progress branch to deploy Taproot and related rules to Bitcoin Core.

The reference implementation has only begun review and the BIPs themselves are subject to [changes](https://github.com/bitcoin/bips/pull/893) which will have to be applied in the code. Some updates may require changes in bcrypto. ~(In fact, bcrypto currently only offers BIP340 Schnorr support in JavaScript, not native code).~ At this time, bcrypto has already made the changes for even-Y instead of square-Y. When the reference branch gets updated I can regenerate the test vectors and rebase.

As such, this PR will require long-term maintenance and paying close attention to the reference implementation, including hard-to-spot [force-push updates](https://github.com/bitcoin/bitcoin/compare/1f499c5b67e9f42e90ac115b43b7c7aa9b44a75c..9f578bfcc7f0b63b50371f5ae046edf2e98fda5b#diff-d22bc3e058f8982972e2eb381a1df668).

There are rules in the code that are not documented in the BIPs (specifically, standardness rules).

There are rules in the code that are not covered by tests. I plan on writing tests, if I can, for the Bitcoin Core branch, then adding them back to this PR if they are considered appropriate. From what I can tell so far the untested rules include some standardenss changes, taproot-embedded-in-P2SH (which is not Taproot and therefore just an anyone-can-spend), and the use of `OP_CHECKSIGADD` with more than one key.

The test cases for this PR were generated by a [modified branch](https://github.com/pinheadmz/bitcoin/tree/taproottest1) of the reference implementation. The Bitcoin Test Framework generates (in memory) a huge amount of transactions using random values and conditions to cover the new rules. The modified branch dumps all the UTXOs and spending transactions into a [JSON file](https://github.com/pinheadmz/bitcoin/blob/taproottest1/taproot_tx_data_single_input.json) along with metadata so that bcoin can be tested the same way. The Bitcoin Test Framework has two sets of tests: a round of single-output spends, and a round of multiple-output spends. The former set creates a somewhat small (635 kB) JSON file, and that is included in this branch as a test vector. The latter creates a huge JSON file - almost 16 MB. I have generated this file and tested bcoin against it locally (`taproot-test.js` can run on either file) but decided not to commit a 16MB test vector. The result of running the multiple-output test is a fabulous report of `18010 passing`.

simongfxu commented:
# [Codecov](https://codecov.io/gh/bcoin-org/bcoin/pull/948?src=pr&el=h1) Report
> Merging [#948](https://codecov.io/gh/bcoin-org/bcoin/pull/948?src=pr&el=desc) into [master](https://codecov.io/gh/bcoin-org/bcoin/commit/1e920cb39830113c239ef1ce5cb02482613de246&el=desc) will **increase** coverage by `0.46%`.
> The diff coverage is `n/a`.

[![Impacted file tree graph](https://codecov.io/gh/bcoin-org/bcoin/pull/948/graphs/tree.svg?width=650&height=150&src=pr&token=C2YIkCZjiU)](https://codecov.io/gh/bcoin-org/bcoin/pull/948?src=pr&el=tree)

```diff
@@            Coverage Diff             @@
##           master     #948      +/-   ##
==========================================
+ Coverage   62.10%   62.57%   +0.46%     
==========================================
  Files         156      157       +1     
  Lines       26076    26426     +350     
==========================================
+ Hits        16194    16535     +341     
- Misses       9882     9891       +9     
```


| [Impacted Files](https://codecov.io/gh/bcoin-org/bcoin/pull/948?src=pr&el=tree) | Coverage Δ | |
|---|---|---|
| [lib/script/witness.js](https://codecov.io/gh/bcoin-org/bcoin/pull/948/diff?src=pr&el=tree#diff-bGliL3NjcmlwdC93aXRuZXNzLmpz) | `66.43% <0.00%> (+4.96%)` | :arrow_up: |
| [lib/utils/taggedhash.js](https://codecov.io/gh/bcoin-org/bcoin/pull/948/diff?src=pr&el=tree#diff-bGliL3V0aWxzL3RhZ2dlZGhhc2guanM=) | `100.00% <0.00%> (ø)` | |
| [lib/script/script.js](https://codecov.io/gh/bcoin-org/bcoin/pull/948/diff?src=pr&el=tree#diff-bGliL3NjcmlwdC9zY3JpcHQuanM=) | `81.70% <0.00%> (+1.73%)` | :arrow_up: |
| [lib/primitives/tx.js](https://codecov.io/gh/bcoin-org/bcoin/pull/948/diff?src=pr&el=tree#diff-bGliL3ByaW1pdGl2ZXMvdHguanM=) | `85.34% <0.00%> (+2.06%)` | :arrow_up: |
| [lib/primitives/coin.js](https://codecov.io/gh/bcoin-org/bcoin/pull/948/diff?src=pr&el=tree#diff-bGliL3ByaW1pdGl2ZXMvY29pbi5qcw==) | `68.29% <0.00%> (+4.06%)` | :arrow_up: |

------

[Continue to review full report at Codecov](https://codecov.io/gh/bcoin-org/bcoin/pull/948?src=pr&el=continue).
> **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta)
> `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> Powered by [Codecov](https://codecov.io/gh/bcoin-org/bcoin/pull/948?src=pr&el=footer). Last update [1e920cb...19c059d](https://codecov.io/gh/bcoin-org/bcoin/pull/948?src=pr&el=lastupdated). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments).

tynes commented on lib/script/common.js:
```diff
@@ -466,7 +483,31 @@ exports.typesByVal = {
+  0: 'LEGACY',
+  1: 'WITNESSV0',
+  2: 'TAPROOT'
 };
```
Missing tapscript here?

pinheadmz replied:
🙏

pinheadmz commented:
rebase to 4e1a95390d513863ee82d1280c989d7a8452586f:

Squash fixup commit for `isTaprootHashType()`

pinheadmz commented:
rebase to 96eb409ba94f32a6bf1f77b513855ed87bf42ceb:

remove JSONify witness stack (helpful for testing, see https://github.com/bcoin-org/bcoin/pull/895)

squash updated test vectors

pinheadmz commented:
It's alive!! Big rebase to ee27f2e5d55e03888de43ebc6f681c9b9ab4285b

- Rebase on current master, which includes latest `bcrypto` package that has already been updated for BIP340 Schnorr final draft with 32-byte public keys and ["even R" tiebreaker](https://github.com/bcoin-org/bcrypto/commit/b9f953c6b5f094571640a55b5dc56f48ee0a457a)
- All new test vectors exported from `feature_taproot.py` in a [modified Bitcoin Core v0.21.1](https://github.com/pinheadmz/bitcoin/tree/taproottest-0.21.1)
- BIP341 TX message (sighash algorithm) has been updated since this PR was originally opened: https://github.com/bitcoin/bips/pull/920
- `SCRIPT_VERIFY_DISCOURAGE_UNKNOWN_ANNEX` flag has been removed since this PR was originally opened, annex is now checked (and rejected) IsWitnessStandard: https://github.com/bitcoin/bitcoin/pull/17977#discussion_r476166522

The new test vectors uncovered some bugs from the original PR:
- Since BIP341 sighashing uses all single SHA256 (and legacy, witnessV0 sighashing uses double-SHA256) the caches for these different operations must be kept separate. You can see the Bitcoin Core version of this [here.](https://github.com/bitcoin/bitcoin/blob/b0e16eb3ac93a8f92e4a9eb7574e539cb5da2335/src/script/interpreter.cpp#L1430-L1446)

Since the tests are still not 100% passing, I'm leaving this PR as a draft and will rebase again once I iron out the last few bugs.

pinheadmz commented:
rebase to f3b2927578fb9a6bc95603e2f94aec158fe6b7d5:

Various bug fixes and new comments:
- fixed taproot-test `getTests()` which was combining vectors together and making a mess
- updated certain error codes with names that match Bitcoin Core (like `SCHNORR_SIG_HASHTYPE`)
- fixed which tapscript witness items get checked against policy size limits
- fixed `evalChecksigTapscript()` handling of empty scripts and return values: [explanation](https://gist.github.com/pinheadmz/55a90e454458a630f22cfa4f3893a791)

The testing suite is now completely passing, so this PR is ready for review! 🎉  

TODO:
- [ ] activation logic
- [ ] Address, Witness and Input methods that "guess" their corresponding output should be checked for interference by new Taproot witness patterns, and expanded if possible. New functions in Script/common.js like "isSignatureEncoding()" will need to be added.
